File: src/waldiez/models/Agent/Reasoning/ReasoningAgentData.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2024 - 2025 Waldiez & contributors
 */
import {
    WaldiezAgentCodeExecutionConfig,
    WaldiezAgentData,
    WaldiezAgentHumanInputMode,
    WaldiezAgentLinkedTool,
    WaldiezAgentNestedChat,
    WaldiezAgentTerminationMessageCheck,
    WaldiezAgentUpdateSystemMessage,
} from "@waldiez/models/Agent/Common";
import { WaldiezReasoningAgentReasonConfig } from "@waldiez/models/Agent/Reasoning/types";
import { WaldiezTransitionTarget } from "@waldiez/models/common/Handoff";

/**
 * Default configuration for Waldiez Reasoning Agent.
 * @see {@link WaldiezReasoningAgentReasonConfig}
 */
export const defaultReasonConfig: WaldiezReasoningAgentReasonConfig = {
    method: "beam_search",
    maxDepth: 3,
    forestSize: 1,
    ratingScale: 10,
    beamSize: 3,
    answerApproach: "pool",
    nsim: 3,
    explorationConstant: 1.41,
};

/**
 * Waldiez Reasoning Agent Data.
 * @param humanInputMode - The human input mode of the agent ("NEVER" | "ALWAYS" | "TERMINATE")
 * @param systemMessage - The system message of the agent
 * @param codeExecutionConfig - The code execution configuration of the agent
 * @param agentDefaultAutoReply - The default auto reply of the agent
 * @param maxConsecutiveAutoReply - The maximum consecutive auto reply of the agent
 * @param termination - The termination message check of the agent
 * @param modelIds - The agent's model ids
 * @param tools - The tools available to the agent
 * @param parentId - The parent id of the agent
 * @param nestedChats - The nested chats of the agent
 * @param contextVariables - The context variables of the agent
 * @param updateAgentStateBeforeReply - The update agent state before reply of the agent
 * @param afterWork - The handoff transition after work of the agent
 * @param handoffs - The handoff / edge ids (used for ordering if needed)
 * @param verbose - The verbose flag of the agent
 * @param reasonConfig - The reasoning configuration of the agent
 * @see {@link WaldiezAgentData}
 * @see {@link WaldiezAgentLinkedTool}
 * @see {@link WaldiezAgentNestedChat}
 * @see {@link WaldiezAgentTerminationMessageCheck}
 * @see {@link WaldiezReasoningAgentReasonConfig}
 * @see {@link defaultReasonConfig}
 * @see {@link WaldiezAgentHumanInputMode}
 * @see {@link WaldiezAgentCodeExecutionConfig}
 * @see {@link WaldiezAgentUpdateSystemMessage}
 * @see {@link WaldiezTransitionTarget}
 */
export class WaldiezAgentReasoningData extends WaldiezAgentData {
    verbose: boolean;
    reasonConfig: WaldiezReasoningAgentReasonConfig;

    constructor(
        props: {
            humanInputMode: WaldiezAgentHumanInputMode;
            systemMessage: string | null;
            codeExecutionConfig: WaldiezAgentCodeExecutionConfig;
            agentDefaultAutoReply: string | null;
            maxConsecutiveAutoReply: number | null;
            termination: WaldiezAgentTerminationMessageCheck;
            modelIds: string[];
            tools: WaldiezAgentLinkedTool[];
            parentId?: string | null;
            nestedChats: WaldiezAgentNestedChat[];
            contextVariables: Record<string, any>;
            updateAgentStateBeforeReply: WaldiezAgentUpdateSystemMessage[];
            afterWork: WaldiezTransitionTarget | null;
            handoffs: string[]; // handoff / edge ids
            verbose: boolean;
            reasonConfig: WaldiezReasoningAgentReasonConfig;
        } = {
            humanInputMode: "NEVER",
            systemMessage: null,
            codeExecutionConfig: false,
            agentDefaultAutoReply: null,
            maxConsecutiveAutoReply: null,
            termination: {
                type: "none",
                keywords: [],
                criterion: null,
                methodContent: null,
            },
            modelIds: [],
            tools: [],
            parentId: undefined,
            nestedChats: [
                {
                    messages: [],
                    triggeredBy: [],
                    condition: {
                        conditionType: "string_llm",
                        prompt: "Start a nested chat",
                    },
                    available: {
                        type: "none",
                        value: "",
                    },
                },
            ],
            contextVariables: {},
            updateAgentStateBeforeReply: [],
            handoffs: [],
            afterWork: null,
            verbose: true,
            reasonConfig: defaultReasonConfig,
        },
    ) {
        super(props);
        this.verbose = props.verbose;
        this.reasonConfig = props.reasonConfig;
    }
}

File: waldiez/models/agents/reasoning/reasoning_agent_data.py
# SPDX-License-Identifier: Apache-2.0.
# Copyright (c) 2024 - 2025 Waldiez and contributors.
# pyright: reportArgumentType=false
"""Reasoning agent data model."""

from typing import Any

from pydantic import Field
from typing_extensions import Annotated, Literal

from ..agent import WaldiezAgentData
from .reasoning_agent_reason_config import WaldiezReasoningAgentReasonConfig


class WaldiezReasoningAgentData(WaldiezAgentData):
    """Reasoning agent data model."""

    human_input_mode: Annotated[
        Literal["ALWAYS", "NEVER", "TERMINATE"],
        Field(
            "NEVER",
            title="Human input mode",
            description="The human input mode, Defaults to `NEVER`",
            alias="humanInputMode",
        ),
    ]
    verbose: Annotated[
        bool,
        Field(
            True,
            title="Verbose",
            description="Whether to show intermediate steps",
        ),
    ]
    reason_config: Annotated[
        WaldiezReasoningAgentReasonConfig,
        Field(
            title="Reason config",
            description="The reasoning agent's reason configuration",
            default_factory=WaldiezReasoningAgentReasonConfig,
            alias="reasonConfig",
        ),
    ]

    @property
    def method(self) -> str:
        """Get the method of the reasoning agent.

        Returns
        -------
        str
            The method of the reasoning agent.

        """
        return self.reason_config.method

    @property
    def max_depth(self) -> int:
        """Get the maximum depth of the reasoning agent.

        Returns
        -------
        int
            The maximum depth of the reasoning agent.

        """
        return self.reason_config.max_depth

    @property
    def beam_size(self) -> int:
        """Get the beam size of the reasoning agent.

        Returns
        -------
        int
            The beam size of the reasoning agent.

        """
        return self.reason_config.beam_size

    @property
    def answer_approach(self) -> str:
        """Get the answer approach of the reasoning agent.

        Returns
        -------
        str
            The answer approach of the reasoning agent.

        """
        return self.reason_config.answer_approach

    @property
    def forest_size(self) -> int:
        """Get the forest size of the reasoning agent.

        Returns
        -------
        int
            The forest size of the reasoning agent.

        """
        return self.reason_config.forest_size

    @property
    def rating_scale(self) -> int:
        """Get the rating scale of the reasoning agent.

        Returns
        -------
        int
            The rating scale of the reasoning agent.

        """
        return self.reason_config.rating_scale

    @property
    def exploration_constant(self) -> float:
        """Get the exploration constant of the reasoning agent.

        Returns
        -------
        float
            The exploration constant of the reasoning agent.

        """
        return self.reason_config.exploration_constant

    @property
    def nsim(self) -> int:
        """Get the number of simulations of the reasoning agent.

        Returns
        -------
        int
            The number of simulations of the reasoning agent.

        """
        return self.reason_config.nsim

    def get_reasoning_config(self) -> dict[str, Any]:
        """Get the reasoning configuration based on the reason_config method.

        Returns
        -------
        dict[str, Any]
            The reasoning configuration.

        """
        reason_dict: dict[str, Any] = {
            "method": self.reason_config.method,
            "max_depth": self.reason_config.max_depth,
            "forest_size": self.reason_config.forest_size,
            "rating_scale": self.reason_config.rating_scale,
        }
        if self.reason_config.method == "beam_search":
            reason_dict["beam_size"] = self.reason_config.beam_size
            reason_dict["answer_approach"] = self.reason_config.answer_approach
        if self.reason_config.method in ("mcts", "lats"):
            reason_dict["nsim"] = self.reason_config.nsim
            reason_dict["exploration_constant"] = (
                self.reason_config.exploration_constant
            )
        return reason_dict
